### PR TITLE
[Optimization] Infer predicate under all JoinTypes

### PR DESCRIPTION
**Pull request by JasonLi-cn**
## Which issue does this PR close?

Closes #.

## Rationale for this change

Currently, infer join predicates is supported only when `JoinType` is `Inner`.
https://github.com/apache/datafusion/blob/18b2aaa04d956475ddc74fbbde3725370e2d7bde/datafusion/optimizer/src/push_down_filter.rs#L560-L567

## What changes are included in this PR?

Support for inferring join predicate in other cases except JoinType which is Inner. e.g.

```sql
SELECT * FROM t1 LEFT JOIN t2 ON t1.c0 = t2.c0 WHERE t2.c0 > 8;
```

#### Where Filters

| JoinType  | Infer Direction | Restrict Null |
| :-------: | :-------------: | :-----------: |
|   Inner   |     L <-> R     |       N       |
|   Left    |     L <-> R     |       Y       |
|   Right   |     L <-> R     |       Y       |
| LeftSemi  |     L <-> R     |       Y       |
| RightSemi |     L <-> R     |       Y       |
| LeftAnti  |     L <-> R     |       Y       |
| RightAnti |     L <-> R     |       Y       |
|   Full    |        -        |       -       |

#### On Filters

| JoinType  | Infer Direction | Restrict Null |
| :-------: | :-------------: | :-----------: |
|   Inner   |     L <-> R     |       N       |
|   Left    |     L -> R      |       Y       |
|   Right   |     L <- R      |       Y       |
| LeftSemi  |     L -> R      |       Y       |
| RightSemi |     L -> R      |       Y       |
| LeftAnti  |        -        |       -       |
| RightAnti |        -        |       -       |
|   Full    |        -        |       -       |

##### Infer Direction

- L -> R: predicate of left table can infer predicate of right.
- L <- R: predicate of right table can infer predicate of left.
- L <-> R: bidirectional infer.
- '-' : do not support infer.

##### Restrict Null

Indicates whether the predicate can filter NULLs,
- N: no, e.g. `t1.c0 IS NULL`
- Y: yes, e.g. `t1.c0 > 8`

### Next Plan

Improved `Push Down Predicates Of Join`. Currently, the pushdown logic in `push_down_all_join` is complete only if `JoinType` is `Inner`. Some predicate(witch is restrict null) that can be pushed down is not supported when `JoinType` is of other types. e.g.

```sql
SELECT * FROM t1 LEFT JOIN t2 ON t1.c0 = t2.c0 WHERE t2.c1 > 8;
```

Due to the limitations of the `lr_is_preserved` function, predicate `t2.c1 > 8` can not be pushed down to right side now. 
https://github.com/apache/datafusion/blob/18b2aaa04d956475ddc74fbbde3725370e2d7bde/datafusion/optimizer/src/push_down_filter.rs#L136-L173

Expect

#### Where Filters

| JoinType  | Allow Pushdown | Restrict Null |
| :-------: | :------------: | :-----------: |
|   Inner   |      L, R      |       N       |
|   Left    |      L, R      |       Y       |
|   Right   |      L, R      |       Y       |
| LeftSemi  |      L, R      |       Y       |
| RightSemi |      L, R      |       Y       |
| LeftAnti  |      L, R      |       Y       |
| RightAnti |      L, R      |       Y       |
|   Full    |       -        |       -       |
#### On Filters

| JoinType  | Allow Pushdown | Restrict Null |
| :-------: | :------------: | :-----------: |
|   Inner   |      L, R      |       N       |
|   Left    |       R        |       Y       |
|   Right   |       L        |       Y       |
| LeftSemi  |      L,R       |       Y       |
| RightSemi |      L,R       |       Y       |
| LeftAnti  |       R        |       Y       |
| RightAnti |       L        |       Y       |
|   Full    |       -        |       -       |

## Are these changes tested?

yes

### CODE REVIEW

**Review by Dandandan — APPROVED** [review]
I think this looks good

File: datafusion/optimizer/src/utils.rs [hunk 1]
```diff
@@ -117,3 +124,165 @@ pub fn log_plan(description: &str, plan: &LogicalPlan) {
...
+    let input_batch = RecordBatch::try_new(Arc::new(schema.clone()), vec![column])?;
+    let execution_props = ExecutionProps::default();
+    let null_column = Column::from_name(DUMMY_COL_NAME);
+
+    let join_cols_to_replace = join_cols_of_predicate
+        .into_iter()
+        .map(|column| (column, &null_column))
+        .collect::<HashMap<_, _>>();
+
+    debug!("predicate: {}", &predicate);
```
Comment (comphead) [inline datafusion/optimizer/src/utils.rs:152]: do we need debug stmts?

File: datafusion/optimizer/src/push_down_filter.rs [hunk 2]
```diff
@@ -2907,6 +3064,46 @@ Projection: a, b
         assert_optimized_plan_eq(plan, expected)
     }
 
+    #[test]
```
Comment (comphead) [inline datafusion/optimizer/src/push_down_filter.rs:3067]: Assuming other outer join types(Left, Right, Full) covered by existing tests?

File: datafusion/optimizer/src/push_down_filter.rs [hunk 3]
```diff
@@ -2044,7 +2161,7 @@ mod tests {
         let expected = "\
         Filter: test2.a <= Int64(1)\
         \n  Left Join: Using test.a = test2.a\
-        \n    TableScan: test\
+        \n    TableScan: test, full_filters=[test.a <= Int64(1)]\
```
Comment (alamb) [inline datafusion/optimizer/src/push_down_filter.rs:2164]: 👍

Comment (alamb) [reply datafusion/optimizer/src/push_down_filter.rs:2164]: these plan changes look good to me

**Review by alamb — COMMENTED** [review]
Thanks @JasonLi-cn  and @Dandandan

File: datafusion/optimizer/src/utils.rs (hunk 1 above)
Comment (JasonLi-cn) [reply datafusion/optimizer/src/utils.rs:152]: I use `debug!` to track changes to the predicate during development. Do we need to remove `debug!`?

(I may have misunderstood what you said :sweat_smile:)

File: datafusion/optimizer/src/push_down_filter.rs (hunk 2 above)
Comment (JasonLi-cn) [reply datafusion/optimizer/src/push_down_filter.rs:3067]: I think the following existing tests can validate for this change:
- Left: [`filter_using_left_join`](https://github.com/JasonLi-cn/arrow-datafusion/blob/feature/optimize_infer_join_predicate/datafusion/optimizer/src/push_down_filter.rs#L2134)
- Right: [`filter_using_right_join`](https://github.com/JasonLi-cn/arrow-datafusion/blob/feature/optimize_infer_join_predicate/datafusion/optimizer/src/push_down_filter.rs#L2172)
- Full: [`full_join_on_with_filter`](https://github.com/JasonLi-cn/arrow-datafusion/blob/ab2a4f5b701650589dcca052e05aae77b5ccaab6/datafusion/optimizer/src/push_down_filter.rs#L2497)

File: datafusion/optimizer/src/utils.rs (hunk 1 above)
Comment (Dandandan) [reply datafusion/optimizer/src/utils.rs:152]: Yes it's cleaner to remove them.

Comment (JasonLi-cn) [reply datafusion/optimizer/src/utils.rs]: Thanks @comphead and @Dandandan

**Comment by alamb** [comment]
Thanks again @JasonLi-cn  and @Dandandan

**Pull request merged**